core/src/subgraph/runner.rs

```diff
@@ -246,15 +255,36 @@ where
                 // TODO: move cancel handle to the Context
                 // This will require some code refactor in how the BlockStream is created
                 let block_start = Instant::now();
-                match self
+
+                let mut action = self
                     .handle_stream_event(event, &block_stream_cancel_handle)
                     .await
                     .map(|res| {
                         self.metrics
                             .subgraph
                             .observe_block_processed(block_start.elapsed(), res.block_finished());
                         res
-                    })? {
+                    })?;
+
+                // It is possible that the subgraph was unassigned, but the runner was in
+                // a retry delay state and did not observe the cancel signal.
+                if block_stream_cancel_handle.is_canceled() {
```

Comment:
Based on the current implementation of [CancelGuard](https://github.com/graphprotocol/graph-node/blob/6b48bfda297eb253e9e802eed53452b9dc1b88d2/graph/src/ext/futures.rs#L70), this should not be possible because it is considered canceled on drop and reassign to the same node replaces the canceler in the context that essentially drops the previous one. Unassign, followed by reassign to a different node will just drop the canceller from the context and it will be considered cancelled.

Reply:
ok from that then I understand we are not missing events, rather the event was processed but we just didn't check is_cancelled often enough. Is that correct?

Reply:
Yes, that is correct. There were several paths that skipped the check, and the fix essentially covers all cases